src/combinatorics/young_diagram.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import order.upper_lower
+
+/-!
+# Young diagrams
+
+A Young diagram is a finite set of up-left justified boxes:
+
+[5, 3, 3, 1]
+□□□□□
+□□□
+□□□
+□
+
+Equivalently, a lower set in `ℕ × ℕ` in the product partial order. We write `(i, j) ∈ μ`
+to say that `(i, j)` (in matrix coordinates) is in the Young diagram `μ`.
+
+## Main functions:
+
+  - `young_diagram` : basic definition
+  - `young_diagram.card` : the cardinality (number of boxes)
+
+  - various constructors:
+      `young_diagram.has_empty` : the empty Young diagram
+
+## Notation:
+
+In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). This terminology is used
+below, e.g. in `young_diagram.nw_of`.
+
+## Tags
+
+Young diagram
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+section young_diagram
+
+/-- A Young diagram is a lower set in ℕ × ℕ -/
+@[ext] structure young_diagram :=
+(cells : finset (ℕ × ℕ))
+(is_lower_set : is_lower_set (cells : set (ℕ × ℕ)))
+
+instance young_diagram.set_like : set_like young_diagram (ℕ × ℕ) :=
+{ coe            := λ μ, (μ.cells : set (ℕ × ℕ)),
+  coe_injective' := λ μ ν h, by { rwa [young_diagram.ext_iff, ← finset.coe_inj] } }
+
+@[simp] lemma young_diagram.mem_cells {μ : young_diagram} (c : ℕ × ℕ) :
+  c ∈ μ.cells ↔ c ∈ μ := iff.rfl
+
+/-- In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+    means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). -/
+lemma young_diagram.nw_of (μ : young_diagram) {i1 i2 j1 j2 : ℕ}
+  (hi : i1 ≤ i2) (hj : j1 ≤ j2) (hcell : (i2, j2) ∈ μ) : (i1, j1) ∈ μ :=
+μ.is_lower_set (prod.mk_le_mk.mpr ⟨hi, hj⟩) hcell
+
+instance young_diagram.has_subset : has_subset young_diagram :=
+{ subset := λ μ ν, μ.cells ⊆ ν.cells }
```

Comment:
If you write it like this, then it uses the `set_like` coercion, which has the effect that when you do `intro` to prove the subset relation it will be in the form `p ∈ μ` rather than ``p ∈ μ.cells`.
```suggestion
{ subset := λ μ ν, (μ : set (ℕ × ℕ)) ⊆ ν }
```